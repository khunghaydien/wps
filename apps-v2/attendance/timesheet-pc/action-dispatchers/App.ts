import isNil from 'lodash/isNil';

import {
  catchApiError,
  catchBusinessError,
  withLoading,
} from '@apps/commons/actions/app';
import { getUserSetting } from '@apps/commons/actions/userSetting';
import msg from '@apps/commons/languages';
import { setUserPermission } from '@apps/commons/modules/accessControl/permission';
import { actions as proxyEmployeeInfoActions } from '@apps/commons/modules/proxyEmployeeInfo';
import { actions as standaloneModeActions } from '@apps/commons/modules/standaloneMode';
import UrlUtil from '@apps/commons/utils/UrlUtil';

import EmployeeRepository from '@apps/repositories/EmployeeRepository';

import { Permission } from '@apps/domain/models/access-control/Permission';
import { UserSetting } from '@apps/domain/models/UserSetting';

import { initDailyStampTime } from '@attendance/timesheet-pc/action-dispatchers/StampWidget';

import UseCases from '../UseCases';
import { fetchTimesheetAndAvailableDailyRequest } from './Timesheet';
import { IOutputData as IFetchTimesheetOutputData } from '@attendance/domain/useCases/timesheet/IFetchUseCase';
import Events from '@attendance/timesheet-pc/events';

/**
 * 勤務表画面を初期化する
 */
export const initialize =
  (param: { userPermission: Permission }) =>
  async (dispatch): Promise<void> => {
    const urlQuery = UrlUtil.getUrlQuery();
    const empId = urlQuery && urlQuery.empId;
    const targetDate = urlQuery && urlQuery.targetDate;
    const standalone = (urlQuery && urlQuery.standalone) === '1';
    const isProxy = !isNil(empId);
    if (isProxy && isNil(targetDate)) {
      dispatch(
        catchBusinessError(
          msg().Com_Err_ErrorTitle,
          msg().Com_Err_ProxyPermissionErrorBody,
          msg().Com_Err_ProxyPermissionErrorSolution,
          { isContinuable: false }
        )
      );
    }
    if (standalone) {
      dispatch(standaloneModeActions.enable());
    }

    if (param) {
      dispatch(setUserPermission(param.userPermission));
    }

    await dispatch(
      withLoading(async (): Promise<void> => {
        try {
          // @ts-ignore
          // Cast (void | UserSetting) to UserSetting
          const [userSetting, fetchedTimesheet, _]: [
            UserSetting,
            IFetchTimesheetOutputData,
            void
          ] = await Promise.all([
            dispatch(getUserSetting()),
            UseCases().fetchTimesheet({
              targetDate,
              employeeId: empId,
            }),
            dispatch(async () => {
              if (!isProxy) {
                await dispatch(initDailyStampTime());
              }
            }),
          ]);

          const { timesheet } = fetchedTimesheet;

          const isSuccess = !isNil(timesheet);

          if (!isSuccess) {
            return;
          }

          if (timesheet.isMigratedSummary) {
            return;
          }

          const isCurrentEmployee =
            isNil(empId) || userSetting.employeeId === empId;

          if (!isCurrentEmployee) {
            const [targetEmployee] = await EmployeeRepository.search({
              id: empId || undefined,
              targetDate: targetDate || undefined,
            });
            dispatch(
              proxyEmployeeInfoActions.set({
                id: targetEmployee.id,
                employeeCode: targetEmployee.code,
                employeeName: targetEmployee.name,
                employeePhotoUrl: targetEmployee.user.photoUrl,
                departmentCode: '', // Not used?
                departmentName: targetEmployee.department
                  ? targetEmployee.department.name
                  : '',
                title: targetEmployee.title,
                managerName: targetEmployee.manager
                  ? targetEmployee.manager.name
                  : '',
              })
            );
          }
          Events.initialized.publish({
            fetchedUserSetting: userSetting,
            fetchedTimesheet,
          });
        } catch (e) {
          dispatch(catchApiError(e, { isContinuable: false }));
        }
      })
    );
  };

/**
 * @param {String} targetDate the target date
 * @param {Object} targetEmployee the target employee
 */
export const switchProxyEmployee =
  (targetDate: string, targetEmployee: { id: string }) =>
  async (dispatch): Promise<boolean> => {
    try {
      await dispatch(
        withLoading(async (): Promise<void> => {
          await UseCases().fetchTimesheet({
            targetDate,
            employeeId: targetEmployee.id,
          });
          dispatch(proxyEmployeeInfoActions.set(targetEmployee));
        })
      );
      return true;
    } catch (e) {
      if (e && e.errorCode === 'ATT_DAILY_IMPORT_REGISTERING') {
        dispatch(
          catchBusinessError(
            msg().Com_Lbl_Error,
            msg().Att_Msg_DailyImportProcessing,
            null,
            { isContinuable: true }
          )
        );
      } else {
        dispatch(
          catchBusinessError(
            msg().Com_Lbl_Error,
            msg().Com_Err_ProxyPermissionErrorBody,
            msg().Com_Err_ProxyPermissionErrorSolution,
            { isContinuable: true }
          )
        );
        return false;
      }
    }
  };

/**
 * 承認者を変更した後に実行する
 * 将来的にEventService.publish() に置換する。
 */
export const onChangeApproverEmployee =
  (
    targetDate: string | null | undefined = null,
    endDate: string | null | undefined = null
  ) =>
  async (dispatch) =>
    dispatch(
      withLoading(async () => {
        await dispatch(
          fetchTimesheetAndAvailableDailyRequest(targetDate, endDate)
        );
      })
    );

/**
 * 打刻した後に呼び出す一連のアクション
 */
export const onStampSuccess = () => (_) => {
  Events.stampedTime.publish();
};

export default initialize;
