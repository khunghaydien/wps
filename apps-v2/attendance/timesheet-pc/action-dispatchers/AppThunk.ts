import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { State } from '../modules';

export type AppAction<ReturnType = unknown> = ThunkAction<
  ReturnType,
  State,
  void,
  Action<unknown>
>;

export type AppDispatch = ThunkDispatch<
  State,
  void,
  Action<unknown | AppAction>
>;
