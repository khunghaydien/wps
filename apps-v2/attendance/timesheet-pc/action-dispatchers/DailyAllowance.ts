import { bindActionCreators } from 'redux';

import {
  catchApiError,
  loadingEnd,
  loadingStart,
} from '../../../commons/actions/app';

import Repository from '@attendance/repositories/AttDailyAllowanceRecordRepository';
import { Timesheet as TimesheetFromRemote } from '@attendance/repositories/models/Timesheet';

import { User } from '../../../domain/models/User';

import { actions as dailyAllowanceEntitiesActions } from '../modules/entities/dailyAllowance';
import { actions as dailyAllowanceUiActions } from '../modules/ui/dailyAllowance';

import { CloseEvent } from '../../daily-allowance';
import { AppDispatch } from './AppThunk';

export const loadDailyAllowanceRecords =
  (timesheet: TimesheetFromRemote, empId?: string) =>
  async (dispatch: AppDispatch) => {
    try {
      dispatch(dailyAllowanceEntitiesActions.clear());
      dispatch(dailyAllowanceUiActions.startLoading());
      const dailyAllowanceSummary = await Repository.search({
        empId,
        startDate: timesheet.startDate,
        endDate: timesheet.endDate,
      });
      dispatch(
        dailyAllowanceEntitiesActions.fetchSuccess(
          dailyAllowanceSummary,
          timesheet
        )
      );
    } catch (e) {
      dispatch(catchApiError(e));
    } finally {
      dispatch(dailyAllowanceUiActions.endLoading());
    }
  };

export const loadDailyAllowanceRecordsBetweenPeriods =
  (period: { startDate: string; endDate: string }, empId?: string, onClose?) =>
  async (dispatch: AppDispatch) => {
    const AppService = bindActionCreators(
      { catchApiError, loadingStart, loadingEnd },
      dispatch
    );
    try {
      AppService.loadingStart();
      const dailyAllowanceSummary = await Repository.search({
        empId,
        ...period,
      });
      dispatch(
        dailyAllowanceEntitiesActions.updateRecords(dailyAllowanceSummary)
      );

      onClose({
        dismissed: true,
        saved: false,
        timestamp: false,
        targetDate: period.startDate,
      });
    } catch (e) {
      dispatch(catchApiError(e));
    } finally {
      AppService.loadingEnd();
    }
  };
// eslint-disable-next-line import/prefer-default-export
export const closeDailyAllowance =
  (e: CloseEvent, user?: User, onClose?) => async (dispatch: AppDispatch) => {
    const AppService = bindActionCreators(
      { catchApiError, loadingStart, loadingEnd },
      dispatch
    );

    const promises = [];
    if (e.saved) {
      promises.push(
        dispatch(
          refreshTimesheet(e.targetDate, user ? user.id : undefined, onClose)
        )
      );
    }

    AppService.loadingStart();
    try {
      await Promise.all(promises);
      // eslint-disable-next-line no-shadow
    } catch (e) {
      AppService.catchApiError(e);
    } finally {
      AppService.loadingEnd();
    }
  };

const refreshTimesheet =
  (targetDate: string, targetEmployeeId?: string, onClose?) =>
  async (dispatch: AppDispatch) => {
    const empId = targetEmployeeId || undefined;
    dispatch(
      loadDailyAllowanceRecordsBetweenPeriods(
        { startDate: targetDate, endDate: targetDate },
        empId,
        onClose
      )
    );
  };
