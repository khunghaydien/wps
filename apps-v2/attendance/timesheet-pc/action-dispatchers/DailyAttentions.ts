import { actions as DailyAttentionsActions } from '../modules/ui/dailyAttentions';

import { AppDispatch } from './AppThunk';

/**
 * 警告ダイアログを開きます。
 * @param {Array<string>} messages 表示したいメッセージ一覧
 */
export const showDailyAttentionsDialog =
  (messages: Array<string>) => (dispatch: AppDispatch) => {
    dispatch(DailyAttentionsActions.set(messages));
  };

/**
 * 警告ダイアログを閉じます。
 */
export const hideDailyAttentionsDialog = () => (dispatch: AppDispatch) => {
  dispatch(DailyAttentionsActions.unset());
};
