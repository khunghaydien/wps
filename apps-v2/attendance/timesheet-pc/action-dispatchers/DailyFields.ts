import { bindActionCreators } from 'redux';

import {
  DailyRecordDisplayFieldLayoutItem,
  LAYOUT_ITEM_TYPE,
  LayoutItemNumberValue,
} from '@attendance/domain/models/DailyRecordDisplayFieldLayout';
import {
  ACTIONS_FOR_FIX,
  FixDailyRequest,
} from '@attendance/domain/models/FixDailyRequest';

import {
  actions,
  DailyRecordDisplayFieldLayoutItemValueForUI,
  UpdateValue,
} from '@apps/attendance/timesheet-pc/modules/ui/dailyRecordDisplayFieldLayout';

import { AppDispatch } from './AppThunk';
import { IInputData as ISubmitFixRequestInputData } from '@attendance/domain/useCases/fixDailyRequest/ISubmitUseCase';
import Events from '@attendance/timesheet-pc/events';
import UseCases from '@attendance/timesheet-pc/UseCases';

interface DailyFieldsService {
  update: (param: UpdateValue) => void;
  save: (param: {
    recordId: string;
    row: DailyRecordDisplayFieldLayoutItem[];
    rowValues: {
      [itemId: string]: DailyRecordDisplayFieldLayoutItemValueForUI;
    };
  }) => Promise<void>;
  fixDaily: (
    record: ISubmitFixRequestInputData,
    changed: boolean,
    row: DailyRecordDisplayFieldLayoutItem[],
    rowValues: {
      [itemId: string]: DailyRecordDisplayFieldLayoutItemValueForUI;
    }
  ) => Promise<void>;
  cancelRequest: (fixDailyRequest: FixDailyRequest) => Promise<void>;
}

export default (dispatch: AppDispatch): DailyFieldsService => {
  const fieldsService = bindActionCreators(actions, dispatch);
  return {
    update: fieldsService.updateField,
    save: async (param) => {
      const { recordId, row, rowValues } = param;
      const values = convertToSavedValues(row, rowValues);

      await UseCases().saveFields({ recordId, values });
      Events.updatedDailyRecord.publish();
    },
    fixDaily: async (record, changed, row, rowValues) => {
      let submit = true;
      if (changed) {
        const values = convertToSavedValues(row, rowValues);
        submit = await UseCases()
          .saveFields({ recordId: record.id, values })
          .then(() => true)
          .catch(() => false);
      }

      if (submit) {
        await UseCases().submitFixDailyRequest(record);
      }
      Events.updatedDailyRecord.publish();
    },
    cancelRequest: async (fixDailyRequest) => {
      if (
        fixDailyRequest?.performableActionForFix ===
        ACTIONS_FOR_FIX.CancelRequest
      ) {
        await UseCases().cancelSubmittedFixDailyRequest(fixDailyRequest.id);
      } else if (
        fixDailyRequest?.performableActionForFix ===
        ACTIONS_FOR_FIX.CancelApproval
      ) {
        await UseCases().cancelApprovalFixDailyRequest(fixDailyRequest.id);
      }
      Events.updatedDailyRecord.publish();
    },
  };
};

const convertToSavedValues = (
  row: DailyRecordDisplayFieldLayoutItem[],
  rowValues: {
    [itemId: string]: DailyRecordDisplayFieldLayoutItemValueForUI;
  }
) => {
  return row
    .filter(
      (item) =>
        rowValues[item.id].existing === true &&
        rowValues[item.id].field?.editable
    )
    .map((item) => {
      const fieldValue = rowValues[item.id].value;
      let value = fieldValue.value as string;

      if (item.type === LAYOUT_ITEM_TYPE.NUMBER) {
        const { textValue } = fieldValue as LayoutItemNumberValue;
        value = textValue === '' ? null : textValue;
      }

      return {
        objectName: item.objectName,
        objectItemName: item.objectItemName,
        value,
      };
    });
};
