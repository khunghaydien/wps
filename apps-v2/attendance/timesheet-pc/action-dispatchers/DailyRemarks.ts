import {
  catchApiError,
  loadingEnd,
  loadingStart,
} from '../../../commons/actions/app';

import DailyRecordRepository from '@attendance/repositories/DailyRecordRepository';

import AttRecord from '../models/AttRecord';
import { DailyRemarks } from '@attendance/domain/models/AttDailyRecord';

import { actions as editingDailyRemarksActions } from '../modules/ui/editingDailyRemarks';

import UseCases from '../UseCases';
import { AppDispatch } from './AppThunk';

/**
 * 日次の備考ダイアログを表示する
 */
export const showDialog = (attRecord: AttRecord) => (dispatch: AppDispatch) => {
  dispatch(editingDailyRemarksActions.set(attRecord));
};

/**
 * 日次の備考ダイアログを閉じる
 */
export const hideDialog = editingDailyRemarksActions.unset;

/**
 * 日次の備考を保存する
 * @param {DailyRemarks} dailyRemarks
 * @param {String} resultTargetPeriodStartDate 成功時に再取得・表示する対象となる集計期間の起算日
 * @param {?String} [targetEmployeeId=null] The ID of target employee
 */
export const save =
  (
    dailyRemarks: DailyRemarks,
    resultTargetPeriodStartDate: string,
    targetEmployeeId: string = null
  ) =>
  async (dispatch: AppDispatch): Promise<void> => {
    try {
      dispatch(loadingStart());
      await DailyRecordRepository.saveRemarks(dailyRemarks);
      dispatch(editingDailyRemarksActions.unset());
      await UseCases().reloadTimesheetOnly({
        targetDate: resultTargetPeriodStartDate,
        employeeId: targetEmployeeId,
      });
    } catch (err) {
      dispatch(catchApiError(err));
    } finally {
      dispatch(loadingEnd());
    }
  };
