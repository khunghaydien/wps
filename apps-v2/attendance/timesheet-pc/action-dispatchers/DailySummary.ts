import { bindActionCreators } from 'redux';

import { confirmToComplementInsufficientingRestTime } from '../../../commons/action-dispatchers/DailyStampTimeResult';
import {
  catchApiError,
  loadingEnd,
  loadingStart,
} from '../../../commons/actions/app';

import { User } from '../../../domain/models/User';

import { CloseEvent } from '../../../daily-summary';
import UseCases from '../UseCases';
import { AppDispatch } from './AppThunk';
import { loadDailyTimeTrackRecordsBetweenPeriods } from './DailyTimeTrack';
import { initDailyStampTime } from './StampWidget';
import { loadTimeTrackAlerts } from './TimeTrackAlert';

const refreshTimesheet =
  (targetDate: string, targetEmployeeId?: string) =>
  async (dispatch: AppDispatch) => {
    const empId = targetEmployeeId || undefined;
    const { timesheet } = await UseCases().reloadTimesheetOnly({
      targetDate,
      employeeId: targetEmployeeId,
    });

    // Don't await loading time-tracking APIs
    Promise.all([
      dispatch(loadTimeTrackAlerts(timesheet, empId)),
      dispatch(
        loadDailyTimeTrackRecordsBetweenPeriods(
          { startDate: targetDate, endDate: targetDate },
          timesheet,
          empId
        )
      ),
    ]);
  };

// eslint-disable-next-line import/prefer-default-export
export const closeDailySummary =
  (e: CloseEvent, user?: User) => async (dispatch) => {
    const AppService = bindActionCreators(
      { catchApiError, loadingStart, loadingEnd },
      dispatch
    );

    const promises = [];
    if (e.timestamp) {
      if (e.dailyStampTimeResult) {
        await dispatch(
          confirmToComplementInsufficientingRestTime(e.dailyStampTimeResult)
        );
      }

      promises.push(dispatch(initDailyStampTime()));
    }

    if (e.saved) {
      promises.push(
        dispatch(refreshTimesheet(e.targetDate, user ? user.id : undefined))
      );
    }

    AppService.loadingStart();
    try {
      await Promise.all(promises);
      // eslint-disable-next-line no-shadow
    } catch (e) {
      AppService.catchApiError(e);
    } finally {
      AppService.loadingEnd();
    }
  };
