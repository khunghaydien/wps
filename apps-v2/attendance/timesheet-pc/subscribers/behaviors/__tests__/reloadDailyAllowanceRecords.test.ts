import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { loadDailyAllowanceRecords } from '@attendance/timesheet-pc/action-dispatchers/DailyAllowance';

import behavior from '../reloadDailyAllowanceRecords';

jest.mock('@attendance/timesheet-pc/action-dispatchers/DailyAllowance', () => ({
  loadDailyAllowanceRecords: jest.fn((...args) => (dispatch) => {
    dispatch({ type: 'LOAD_DAILY_ALLOWANCE_RECORDS', payload: args });
    return Promise.resolve();
  }),
}));

const $loadDailyAllowanceRecords = loadDailyAllowanceRecords as jest.Mock;

const createStore = configureMockStore([thunk]);

const defaultState = {
  common: {
    app: {
      error: null,
    },
    userSetting: {
      employeeId: 'employeeId',
    },
  },
};

type Input = Parameters<ReturnType<typeof behavior>>[0];

beforeEach(() => {
  jest.clearAllMocks();
});

it('should do.', async () => {
  // Arrange
  const store = createStore(defaultState);

  // Act
  await behavior(store)({
    employeeId: 'employeeId',
    timesheet: {
      workingTypeList: [
        {
          useAllowanceManagement: true,
        },
      ],
    },
  } as unknown as Input);

  // Assert
  expect($loadDailyAllowanceRecords).toBeCalledTimes(1);
  expect($loadDailyAllowanceRecords).toBeCalledWith(
    {
      workingTypeList: [
        {
          useAllowanceManagement: true,
        },
      ],
    },
    'employeeId'
  );
  expect(store.getActions()).toMatchSnapshot();
});

it.each([null, undefined])(
  'should do if employeeId is %s.',
  async (employeeId) => {
    // Arrange
    const store = createStore(defaultState);

    // Act
    await behavior(store)({
      employeeId,
      timesheet: {
        workingTypeList: [
          {
            useAllowanceManagement: true,
          },
        ],
      },
    } as unknown as Input);

    // Assert
    expect($loadDailyAllowanceRecords).toBeCalledTimes(1);
    expect($loadDailyAllowanceRecords).toBeCalledWith(
      {
        workingTypeList: [
          {
            useAllowanceManagement: true,
          },
        ],
      },
      undefined
    );
    expect(store.getActions()).toMatchSnapshot();
  }
);

it('should not do if timesheet is null', async () => {
  // Arrange
  const store = createStore(defaultState);

  // Act
  await behavior(store)({ timesheet: null } as unknown as Input);

  // Assert
  expect($loadDailyAllowanceRecords).toBeCalledTimes(0);
  expect(store.getActions()).toMatchSnapshot();
});

it('should not do if timesheet is migrated', async () => {
  // Arrange
  const store = createStore(defaultState);

  // Act
  await behavior(store)({
    timesheet: {
      isMigratedSummary: true,
    },
  } as unknown as Input);

  // Assert
  expect($loadDailyAllowanceRecords).toBeCalledTimes(0);
  expect(store.getActions()).toMatchSnapshot();
});
