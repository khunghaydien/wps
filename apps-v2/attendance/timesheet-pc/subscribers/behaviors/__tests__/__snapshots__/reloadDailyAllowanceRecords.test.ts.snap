// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should do if employeeId is null. 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "workingTypeList": Array [
          Object {
            "useAllowanceManagement": true,
          },
        ],
      },
      undefined,
    ],
    "type": "LOAD_DAILY_ALLOWANCE_RECORDS",
  },
]
`;

exports[`should do if employeeId is undefined. 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "workingTypeList": Array [
          Object {
            "useAllowanceManagement": true,
          },
        ],
      },
      undefined,
    ],
    "type": "LOAD_DAILY_ALLOWANCE_RECORDS",
  },
]
`;

exports[`should do. 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "workingTypeList": Array [
          Object {
            "useAllowanceManagement": true,
          },
        ],
      },
      "employeeId",
    ],
    "type": "LOAD_DAILY_ALLOWANCE_RECORDS",
  },
]
`;

exports[`should not do if timesheet is migrated 1`] = `Array []`;

exports[`should not do if timesheet is null 1`] = `Array []`;
